packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenService.m

```diff
@@ -86,24 +88,40 @@ - (void)hideSplashScreenFor:(UIViewController *)viewController
   if (![self.splashScreenControllers objectForKey:viewController]) {
     return failureCallback(@"No native splash screen registered for given view controller. Call 'SplashScreen.show' for given view controller first.");
   }
-  return [[self.splashScreenControllers objectForKey:viewController] hideWithCallback:successCallback
-                                                                      failureCallback:failureCallback];
+  [UIApplication.sharedApplication.keyWindow removeObserver:self forKeyPath:kRootViewController context:nil];
+  EXSplashScreenViewController *splashScreenViewController = [self.splashScreenControllers objectForKey:viewController];
+
+  UM_WEAKIFY(self);
+  return [splashScreenViewController
+          hideWithCallback:^(BOOL hasEffect) { UM_ENSURE_STRONGIFY(self); [self.splashScreenControllers removeObjectForKey:viewController]; }
+          failureCallback:^(NSString *message) { UM_ENSURE_STRONGIFY(self); [self.splashScreenControllers removeObjectForKey:viewController]; }];
```

Comment:
I think we may want to call `successCallback(hasEffect)` and `failureCallback(message)` respectively in these new blocks as well? Or if the blocks that are passed in are always empty we should just remove the parameters entirely.

Reply:
good catch! thanks for the help 👍